Lesson_11/Tusk_1.py

```diff
@@ -0,0 +1,20 @@
+from time import ctime
+
+
+def func_decorator(func):
+    def inside_func(*args, **kwargs):
+        print(f'You call the function {func}')
+        func(*args, **kwargs)
```

Comment:
One more thing: assign the result of this call to some variable and return the variable after the last print to be able to decorate functions that return some value

Reply:
Thank you for comment. I did it. 
But to be honest, i didnt understand why. 
I tryed my code without variable and it works even with a lot of args in function. 
Or maybe i get something wrong?

Reply:
Let's say we have a function:
```
def sum_of_two(a, b):
    return a + b
```

Now try to decorate this function with these variants of decorators:
```
def deco_without_args_and_return(func):
    def wrap():
        print("I am a decorator")
        func()
    return wrap
```
Here you will have `TypeError: deco_without_args_and_return.<locals>.wrap() takes 0 positional arguments but 2 were given
`

```
def deco_without_args(func):
    def wrap():
        print("I am a decorator")
        return func()
    return wrap
```
Here you will have
`TypeError: deco_without_args.<locals>.wrap() takes 0 positional arguments but 2 were given`


```
def deco_without_return(func):
    def wrap(*args, **kwargs):
        print("I am a decorator")
        func(*args, **kwargs)
    return wrap
```
Here you will not have an error, but the decorated function will return None instead of sum of two arguments:

```
res = sum_of_two(1, 2)
print(res) # prints None instead of 3 because there is no return inside of the decorator
```
This is happening because technically the real function call we have inside of the decorator: `func(*args, **kwargs)` and if we ignore the result here (don't return it) it will be lost.


```
def deco_with_all(func):
    def wrap(*args, **kwargs):
        print("I am a decorator")
        return func(*args, **kwargs)
    return wrap
```

And only this decorator will work correctly

Reply:
And you need to keep the result in a variable because you have a print after function is called. So, according to my previous comment, to keep the result of the function call, we need to store that result somewhere. In our case, this somewhere is a variable `result`